### PR TITLE
introduce bind-ssh support

### PR DESCRIPTION
**Pull request by cjp256**
Introduce support for binding user's .ssh to local LXD and Multipass build environments.

Depends on #2786 and #2078

### CODE REVIEW

**Comment by cjp256** [comment]
Should we mark this as experimental?

File: snapcraft/internal/build_providers/_base_provider.py [hunk 1]
```diff
@@ -126,6 +127,7 @@ def __init__(
         if environment_passthrough is None:
             environment_passthrough = dict()
         self.environment_passthrough = environment_passthrough
+        self._cached_home_directory: pathlib.Path = None
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_base_provider.py]: :100:

**Comment by sergiusens** [comment]
Can you split 9c77cb69e8ef368327963f76f38078d99881abddin two? One without changing existing signatures and another with signature changes/concentration of code?

**Comment by sergiusens** [comment]
It feels like we need to rename this base provider to LocalBaseProvider (to make room for the remote one) considering moving most of the only local things into this class.

**Comment by sergiusens** [comment]
Does this indeed work? It would be interesting to see (albeit might not be possible today), to mount this path read only

**Comment by cjp256** [comment]
> Can you split [9c77cb6](https://github.com/snapcore/snapcraft/commit/9c77cb69e8ef368327963f76f38078d99881abdd) two? One without changing existing signatures and another with signature changes/concentration of code?

Absolutely

**Comment by cjp256** [comment]
> Does this indeed work? It would be interesting to see (albeit might not be possible today), to mount this path read only

Yeah I tested in LXD and multipass.  A read-only option would be good, but might cause grief with known_hosts...

**Comment by cjp256** [comment]
Requires https://github.com/snapcore/snapcraft/pull/2678

**Comment by cjp256** [comment]
Depends on #2786.

File: snapcraft/internal/build_providers/_base_provider.py [hunk 2]
```diff
@@ -190,16 +193,39 @@ def _push_file(self, *, source: str, destination: str) -> None:
         """Push a file into the instance."""
 
     @abc.abstractmethod
+    def is_mounted(self, target: str) -> bool:
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_base_provider.py]: the model we decided to follow for this base class is that if we do not intend to use these methods outside of the components inheriting from it, they should be private/protected (start with `_`)

File: snapcraft/internal/build_providers/_base_provider.py [hunk 3]
```diff
@@ -190,16 +193,39 @@ def _push_file(self, *, source: str, destination: str) -> None:
...
+        """Query if there is a mount at target mount point."""
+
+    @abc.abstractmethod
+    def mount(self, host_source: str, target: str) -> None:
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_base_provider.py]: the model we decided to follow for this base class is that if we do not intend to use these methods outside of the components inheriting from it, they should be private/protected (start with `_`)

**Comment by cjp256** [comment]
Added commit: https://github.com/snapcore/snapcraft/pull/2791/commits/005119a9bb406215ee946db3f6f501197f6dba26

Force push diff:
https://github.com/snapcore/snapcraft/compare/259bfcf17e95110dbaee955b2aeaebb91c55b6e2..857851bf47e198211be5c474437b2f4eabdf99a1

Should be good to go.

**Pull request merged**